File: aha_algorithms/p98_chapter4_4.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

game_map = """
#############
#GG.GGG#GGG.#
###.#G#G#G#G#
#.......#..G#
#G#.###.#G#G#
#GG.GGG.#.GG#
#G#.#G#.#.#.#
##G...G.....#
#G#.#G###.#G#
#...G#GGG.GG#
#G#.#G#G#.#G#
#GG.GGG#G.GG#
#############"""

game_map = [list(i) for i in game_map.split()]
edge, startx, starty = (13, 3, 3)

class Note(object):
    def __init__(self, x, y, num=0):
        self.x = x
        self.y = y
        self.num = num


def getnum(note_x, note_y):
    _sum = 0
    for i in range(note_x-1, -1, -1):
        if game_map[i][note_y] == '#':
            break
        elif game_map[i][note_y] == 'G':
            _sum += 1

    for i in range(note_x, edge):
        if game_map[i][note_y] == '#':
            break
        elif game_map[i][note_y] == 'G':
            _sum += 1

    for i in range(note_y-1, -1, -1):
        if game_map[note_x][i] == '#':
            break
        elif game_map[note_x][i] == 'G':
            _sum += 1

    for i in range(note_y, edge):
        if game_map[note_x][i] == '#':
            break
        elif game_map[note_x][i] == 'G':
            _sum += 1

    return _sum


def main():
    que = [Note(0, 0) for _ in range(401)]
    _max = Note(0, 0, 0)
    book = [[0 for i in range(20)] for j in range(20)]
    _next = [
        [0, 1],
        [1, 0],
        [0, -1],
        [-1, 0],
    ]
    head = tail = 1
    que[tail] = Note(startx, starty, getnum(startx, starty))
    tail += 1
    book[startx][starty] = 1

    while head<tail:
        for dx, dy in _next:
            tx = que[head].x + dx
            ty = que[head].y + dy
            if tx<0 or tx>edge-1 or ty<0 or ty>edge-1:
                continue
            if game_map[tx][ty] == '.' and book[tx][ty] == 0:
                book[tx][ty] = 1
                que[tail] = Note(tx, ty, getnum(tx, ty))
                if _max.num < que[tail].num:
                    _max = que[tail]
                tail += 1
        head += 1
    print "location (%d, %d), kill %d enemy.\n" % (_max.x, _max.y, _max.num)

main()

File: aha_algorithms/p98_chapter4_4.c
#include <stdio.h>

// 广度优先搜索
struct note{
    int x;
    int y;
};

char a[20][21];

int getnum(int i, int j){
    int sum, x, y;
    sum = 0;
    x = i; y = j;
    while (a[x][y] != '#'){
        if (a[x][y] == 'G'){
            sum++;
        }
        x--;
    }

    x = i; y = j;
    while (a[x][y] != '#'){
        if (a[x][y] == 'G'){
            sum++;
        }
        x++;
    }

    x = i; y = j;
    while (a[x][y] != '#'){
        if (a[x][y] == 'G'){
            sum++;
        }
        y--;
    }

    x = i; y = j;
    while (a[x][y] != '#'){
        if (a[x][y] == 'G'){
            sum++;
        }
        y++;
    }

    return sum;
}

int main() {
    struct note que[401];
    int head, tail;
    int book[20][20] = {0};
    int i, j, k, sum, max=0, mx, my, n, m, startx, starty, tx, ty;
    int next[4][2] = {
            {0, 1},
            {1, 0},
            {0, -1},
            {-1, 0},
    };
    scanf("%d %d %d %d", &n, &m, &startx, &starty);
    for (i=0; i<=n-1; i++) {
        scanf("%s", a[i]);
    }
    head = 1;
    tail = 1;
    que[tail].x = startx;
    que[tail].y = starty;
    tail++;
    book[startx][starty] = 1;
    max = getnum(startx, starty);
    mx = startx;
    my = starty;

    while (head<tail) {
        for (k = 0; k <= 3; k++) {
            tx = que[head].x + next[k][0];
            ty = que[head].y + next[k][1];
            if (tx<0 || tx>n-1 || ty<0 || ty>m-1) {
                continue;
            }
            if (a[tx][ty] == '.' && book[tx][ty] == 0) {
                book[tx][ty] = 1;
                que[tail].x = tx;
                que[tail].y = ty;
                tail++;
                sum = getnum(tx, ty);
                if (sum > max) {
                    max = sum;
                    mx = tx;
                    my = ty;
                }
            }
        }
        head++;
    }
    printf("location (%d, %d), kill %d enemy.\n", mx, my, max);
    getchar();
    return 0;
}

/* input
13 13 3 3
#############
#GG.GGG#GGG.#
###.#G#G#G#G#
#.......#..G#
#G#.###.#G#G#
#GG.GGG.#.GG#
#G#.#G#.#.#.#
##G...G.....#
#G#.#G###.#G#
#...G#GGG.GG#
#G#.#G#G#.#G#
#GG.GGG#G.GG#
#############
*/

/* output
location (7, 11), kill 10 enemy.
*/
